moshi/src/main/java/com/squareup/moshi/JsonUtf8Reader.kt

```diff
@@ -0,0 +1,1186 @@
+/*
+ * Copyright (C) 2010 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi
+
+import okio.Buffer
+import okio.BufferedSource
+import okio.ByteString
+import okio.ByteString.Companion.encodeUtf8
+import okio.buffer
+import java.io.EOFException
+import java.io.IOException
+import java.math.BigDecimal
+
+internal class JsonUtf8Reader : JsonReader {
+  /** The input JSON.  */
+  private val source: BufferedSource
+  private val buffer: Buffer
+  private var peeked = PEEKED_NONE
+
+  /**
+   * A peeked value that was composed entirely of digits with an optional leading dash. Positive
+   * values may not have a leading 0.
+   */
+  private var peekedLong: Long = 0
+
+  /** The number of characters in a peeked number literal.  */
+  private var peekedNumberLength = 0
+
+  /**
+   * A peeked string that should be parsed on the next double, long or string. This is populated
+   * before a numeric value is parsed and used if that parsing fails.
+   */
+  private var peekedString: String? = null
+
+  /**
+   * If non-null, the most recent value read was [.readJsonValue]. The caller may be
+   * mid-stream so it is necessary to call [JsonValueSource.discard] to get to the end of the
+   * current JSON value before proceeding.
+   */
+  private var valueSource: JsonValueSource? = null
+
+  constructor(source: BufferedSource?) {
+    if (source == null) {
+      throw NullPointerException("source == null")
+    }
```

Comment:
```suggestion
  constructor(source: BufferedSource) {
```

Reply:
🤦🏾 clearly I missed some stuff, thank you